README.md
# desktop chatai plus

[![GitHub](https://img.shields.io/badge/Github-Repo-blue?logo=github)](https://github.com/danger-dream/desktop-chatai-plus)

desktop chatai plus 是一个高效、轻量级的桌面版AI工具，为用户提供了一个统一、便捷的界面，实现多个LLM平台的即时交互。

## 功能特性

- **常驻内存 & 失焦隐藏 & 快捷键唤起**：保持程序在后台运行，随时恢复，高效响应。
- **简洁界面**：基于Tailwind CSS设计的清晰、简洁的用户界面。
- **多LLM交互**：在同一对话框中与多个LLM平台进行交互。
- **命令式模板**：使用\"/\"前缀快速访问常用功能，如使用自定义的Prompt、切换模型、清空对话等..
- **对话直达平台**：与LLM平台直接交互，无需经过任何第三方服务器。
- **数据保障**：对话数据仅保存在本地，保障用户隐私。
- **即将到来的功能**：web检索、图文对话、文档问答、图片生成、语音交互等。

## 截图
![7.png](images/7.png)
![1.png](./images/1.png)
![2.png](images/2.png)
![3.png](images/3.png)
![4.png](images/4.png)
![5.png](images/5.png)
![6.png](images/6.png)
![so.png](images/so.png)
![sw.png](images/sw.png)
![sq.png](images/sq.png)
![ss.png](images/ss.png)


## 项目背景
在使用LLM的过程中，各家平台有不同的侧重点。 每次切换不同网页也比较麻烦，且经常遇到需要重新登录的情况。虽然社区有人开发过类似的项目，但大多是web项目。开发desktop chatai plus的主要目的是常驻内存，需要时立马打开并辅助解决问题。


## 源码运行

```bash
# 1. 克隆项目到本地
git clone https://github.com/danger-dream/desktop-chatai-plus.git

# 2. 安装依赖
cd desktop-chatai-plus

npm install

# 3. 运行
npm run dev
```

## 啰嗦的..
- 没做过多的组件拆分，感觉没必要，就这么点东西
- 放弃了UI库，前端真不熟，用了看着又不满意
- 依赖做到极简了，markdown那堆插件和tabler-icon会想办法去掉


CHANGELOG
# CHANGELOG

All notable changes to this project will be documented in this file.

## [0.2.5] - 2023-11-10

### Changed
- 允许OpenAI设置MaxTokens为0，即不限制生成长度
- 允许最大关联消息数上限至100条

### Fixed
- 修复Popover无法显示的问题

## [0.2.4] - 2023-11-8

### Changed
- 向自定义命令中传递更多参数、方法，方便在其中进行更复杂的定制
- 新增"系统提示"功能及相关配置
- 新增"修改系统提示"命令，可修改系统提示的内容

### Fixed
- 修复删除对话后未删除关联的音频文件问题

## [0.2.3] - 2023-11-7

### Changed
- 新增OpenAI最新模型支持：gpt-3.5-turbo-1106
- 增加OpenAI TTS模型语音朗读回复功能及相关设置

### Fixed
- 修复无法保存配置的问题
- 修复无法弹出MessageBox的问题
- 修复使用"分割对话"命令后的当前对话无法关联的问题

## [0.2.2] - 2023-11-7

### Changed
- 新增OpenAI最新模型支持：gpt-4-1106-preview、gpt-4-vision-preview
- 切换默认模型为OpenAI: gpt-4-1106-preview


## [0.2.1] - 2023-11-6

### Changed
- 新增Icon组件，为不增加额外依赖，使用组件生成方式创建svg
- 优化popver组件

### Fixed
- 修复输入框中按下Shift + Enter时不换行的问题，修正placeholder内容

### Removed
- Removed @tabler/icons-vue
- Removed markdown-it-*

## [0.2.0] - 2023-11-5

### Added
- 新增"分割对话"命令
- 封装平台代码，使用electron ipc替换，方便移植到tauri

### Changed
- Update electron 19.1.9 to 27.0.3
- Update electron-builder 23.6.0 to 24.6.4

### Removed
- Removed vite-plugin-electron-renderer
- Removed @electron/remote


src/platform/OpenAI.ts
import { IPlatformConfig, ILLMOptions, Message, BasePlatform, IEmbeddingResult, ChunkFunction, ILLMResult, PlatformKeys } from '../../types'
import { getEncoding } from 'js-tiktoken'
const enc = getEncoding('cl100k_base')

export default class OpenAI implements BasePlatform {
	public static readonly label = 'OpenAI'
	public static readonly platformName: PlatformKeys = 'openai'
	public static readonly models = [ 'gpt-4-1106-preview', 'gpt-4-vision-preview', 'gpt-4', 'gpt-4-32k', 'gpt-3.5-turbo', 'gpt-3.5-turbo-1106', 'gpt-3.5-turbo-16k' ]
	private readonly apiKey: string
	private readonly basePath: string
	
	constructor(config: IPlatformConfig) {
		const {apiKey, basePath} = config
		if (!apiKey) {
			throw new Error('OpenAI API key is required')
		}
		this.apiKey = apiKey
		this.basePath = basePath || 'https://api.openai.com/v1'
	}
	
	public getModels(): string[]{
		return OpenAI.models
	}
	
	async embedTexts(texts: string[] = []): Promise<IEmbeddingResult> {
		if (texts.length < 1) {
			throw new Error('text is required');
		}
		try {
			const inputs = texts.map(x => x.replaceAll('\n', ''));
			const response = await fetch(`${ this.basePath }/embeddings`, {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
					'Authorization': `Bearer ${ this.apiKey }`
				},
				body: JSON.stringify({input: inputs, model: 'text-embedding-ada-002'})
			});
			if (!response.ok) {
				return {embeddings: [], total_tokens: 0}
			}
			const result = await response.json();
			return {
				embeddings: result.data.map((x: { embedding: number[][] }) => x.embedding),
				total_tokens: result.usage.total_tokens
			};
		} catch {
			return {embeddings: [], total_tokens: 0};
		}
	}
	
	
	async completion(messages: Message[] | Message | string, options?: ILLMOptions, chunk_callback?: ChunkFunction): Promise<ILLMResult> {
		if (!messages || (Array.isArray(messages) && messages.length < 1)) {
			throw new Error('message is required')
		}
		if (typeof messages === 'string') {
			messages = [ {role: 'user', content: messages} ]
		}
		messages = Array.isArray(messages) ? messages : [ messages ]
		const body = {
			stream: true,
			model: options.model,
			temperature: options.temperature ?? 0.95,
			frequency_penalty: options.penalty_score ?? 0,
			messages
		}
		if (!body.model || !OpenAI.models.includes(body.model.toLowerCase())) {
			body.model = OpenAI.models[0]
		}
		let content = ''
		let total_tokens = 0
		try {
			const resp = await fetch(this.basePath + '/chat/completions', {
				headers: {'Content-Type': 'application/json', 'Authorization': 'Bearer ' + this.apiKey},
				method: 'POST', body: JSON.stringify(body)
			})
			const reader = resp.body.pipeThrough(new TextDecoderStream()).getReader()
			let errStr = ''
			let is_stop = false
			while(!is_stop) {
				const res = await reader.read()
				if(res.done) break
				const chunk = res.value.replace(/data: /g, '').split('\n').map(x => x.trim()).filter(Boolean)
				for (let item of chunk) {
					if (is_stop) break
					if (item == '[DONE]') {
						is_stop = true
						break
					}
					try {
						if(errStr) {
							item = errStr + item
							errStr = ''
						}
						const chunk = JSON.parse(item) as { choices?: { delta?: { content?: string } }[] }
						const text = chunk.choices[0]?.delta?.content
						if (!text) continue
						content += text
						total_tokens++
						if (chunk_callback && !chunk_callback(text)) {
							try {
								await reader.cancel()
							} catch {}
							is_stop = true
							break
						}
						
					}catch {
						errStr = item
					}
				}
			}
		} catch {
		}
		return {content, total_tokens}
	}
	
	count_tokens(text: string): number {
		try {
			return enc.encode(text, 'all').length + 6
		} catch {
			return text.length
		}
	}
	
	async speech(text: string, opts?: Record<string, any>): Promise<ArrayBuffer> {
		const model = opts?.model || 'tts-1'
		const voice = opts?.voice || 'alloy'
		const speed = opts?.speed ?? 1
		const resp = await fetch(this.basePath + '/audio/speech', {
			headers: {'Content-Type': 'application/json', 'Authorization': 'Bearer ' + this.apiKey},
			method: 'POST', body: JSON.stringify({input: text, model, voice, speed})
		})
		if (resp.status === 200) {
			return await resp.arrayBuffer()
		}
		throw new Error('speech failed')
	}
	
	getSystemPrompt(): string {
		const dt = new Date()
		return `You are ChatGPT, a large language model trained by OpenAI.
Knowledge ctoff: 2023-04
Current date: ${ dt.getFullYear()}-${ ((dt.getMonth() + 1) + '').padStart(2, '0') }-${ (dt.getDate() + '').padStart(2, '0') }`
	}
}
